### PR TITLE
search: Enable sub-repo filtering

### PR DESCRIPTION
**Pull request by ryanslade**
This change adds sub-repo filtering in the search aggregator. 

It is very similar to the previous PR  submitted by @bobheadxi, https://github.com/sourcegraph/sourcegraph/pull/27189, as after a lot of back and forth we decided that given there was a single place we could add a filter, this was the most non-intrusive way of enabling it in the search path.

I've tested it by manually setting some filtering rules and we plan to add end to end coverage later.

Closes https://github.com/sourcegraph/sourcegraph/issues/28242

### CODE REVIEW

File: cmd/frontend/graphqlbackend/search_results.go [hunk 1]
```diff
@@ -1737,6 +1738,44 @@ func (r *searchResolver) doResults(ctx context.Context, args *search.TextParamet
...
+			Results: e.Results[:0],
+			Stats:   e.Stats,
+		}
+		for _, r := range e.Results {
+			key := r.Key()
+			content := authz.RepoContent{
+				Repo: key.Repo,
+				Path: key.Path,
+			}
+			perms, err := authz.ActorPermissions(ctx, checker, a, content)
```
Comment (tsenart) [inline cmd/frontend/graphqlbackend/search_results.go:1761]: I think we could make this more performant by batching requests to the DB, rather than doing 1 per Match.

File: cmd/frontend/graphqlbackend/search_results.go [hunk 2]
```diff
@@ -1737,6 +1738,44 @@ func (r *searchResolver) doResults(ctx context.Context, args *search.TextParamet
...
+			if err != nil {
+				errs = multierror.Append(errs, err)
+				continue
+			}
+			if perms.Include(authz.Read) {
+				filtered.Results = append(filtered.Results, r)
+			}
+		}
+		// We don't want to spam our logs or return sensitive authz related errors to the
+		// user so we'll return generic error and log something more specific.
```
Comment (tsenart) [inline cmd/frontend/graphqlbackend/search_results.go:1771]: Can we instead just make sure that the errors we return don't contain sensitive data?

**Review by tsenart — APPROVED** [review]
Please add backend integration tests in this PR as well. https://github.com/sourcegraph/sourcegraph/blob/main/dev/gqltest/search_test.go

File: cmd/frontend/graphqlbackend/search_results.go (hunk 1 above)
Comment (camdencheek) [reply cmd/frontend/graphqlbackend/search_results.go:1761]: Is it possible we could do even better by evaluating the permissions in advance, then just checking a file path against the permissions object in the returned closure? I'd be concerned about even a single db request per stream event.

File: cmd/frontend/graphqlbackend/search_results.go [hunk 3]
```diff
@@ -1737,6 +1738,44 @@ func (r *searchResolver) doResults(ctx context.Context, args *search.TextParamet
...
 
+// subRepoFilterFunc returns a filtering function that applies sub-repo permissions
+// to search events. It is possible that a nil function is returned.
+func subRepoFilterFunc(ctx context.Context, checker authz.SubRepoPermissionChecker) run.EventTransformer {
+	if checker == nil || !checker.Enabled() {
+		return nil
+	}
+
+	a := actor.FromContext(ctx)
+	return func(e streaming.SearchEvent) (streaming.SearchEvent, error) {
```
Comment (camdencheek) [inline cmd/frontend/graphqlbackend/search_results.go:1749]: The returned function should probably take `ctx` as an argument rather than closing over a context, otherwise we're listening for events on a context that might be different than the context that this function is actually executed under.

**Review by keegancsmith — COMMENTED** [review]
agree with the current inline comments around perf. IE the filter func needs to be smart.

Any concerns around errors from backends containing path information? I don't think this is something that is currently possible, but just a thought :)

File: cmd/frontend/graphqlbackend/search_results.go (hunk 1 above)
Comment (ryanslade) [reply cmd/frontend/graphqlbackend/search_results.go:1761]: The checker that is passed in does a bunch of caching already. The first request per user will fetch all their sub-repo perms, build the rules and cache them. This is a cache that persists between requests and expires after a short period, currently 10 seconds. What this means is that we expect at most one db request every 10 seconds per frontend instance, per user. 

https://github.com/sourcegraph/sourcegraph/blob/main/internal/authz/sub_repo_perms.go#L221

**Comment by ryanslade** [comment]
@tsenart I'm struggling to figure out how to do the backend integration tests. 

I've tried modifying [this test](https://github.com/sourcegraph/sourcegraph/blob/main/cmd/frontend/graphqlbackend/search_test.go#L30) but the issue is that Zoekt isn't mocked so we get an error like this:

```
RepoUniverseText search failed: unable to load in-cluster configuration, KUBERNETES_SERVICE_HOST and KUBERNETES_SERVICE_PORT must be defined
```

I assume I need to somehow mock out Zoekt itself using a fake searcher but I'm not sure where the best place to do that is? Maybe somehow injecting it into he `searchResolver`?

**Comment by ryanslade** [comment]
I'm working on a test that just uses the resolver directly which should be enough to test that the pieces are hooked up correctly.

File: cmd/frontend/graphqlbackend/search_results.go (hunk 1 above)
Comment (camdencheek) [reply cmd/frontend/graphqlbackend/search_results.go:1877]: Ah, I see. Thanks for clarifying. I was thinking the checker was just a db instance, but this makes sense. 

In that case, I expect performance here to be good enough, but that's assuming the sync overhead of the cache isn't significant. I don't think that's a given in this case, considering that the sync primitives are going to be hit pretty hard since the cache is shared across all requests, and will be hit for every stream event in the process (as opposed to being scoped to the request, like most of our stuff that needs synchronization). 

TL;DR: please do some performance analysis when this rolls out to check that it doesn't adversely affect search speed/latency. This might be an ideal case to use a rollout feature flag on sourcegraph.com so it can easily be enabled/disabled.

**Comment by tsenart** [comment]
> @tsenart I'm struggling to figure out how to do the backend integration tests.

When say backend integration tests, I mean https://github.com/sourcegraph/sourcegraph/blob/main/dev/gqltest/search_test.go

**Comment by ryanslade** [comment]
> > @tsenart I'm struggling to figure out how to do the backend integration tests.
> 
> When say backend integration tests, I mean https://github.com/sourcegraph/sourcegraph/blob/main/dev/gqltest/search_test.go

🤦

**Comment by ryanslade** [comment]
In order to get the backend integration tests to pass I'm going to need to introduce some new API's in our GraphQL layer that extend our explicit permissions API. That way the integration tests can do some permissions setup before running. 

Since that is going to be a bigger change and will mostly be touching other parts of the code I'm going to do it as a separate PR and merge this one.

File: cmd/frontend/graphqlbackend/search_results.go [hunk 4]
```diff
@@ -1926,6 +1926,45 @@ func (r *searchResolver) doResults(ctx context.Context, jobs []run.Job, repoOpti
...
+		errs := &multierror.Error{}
+		filtered := streaming.SearchEvent{
+			Results: e.Results[:0],
+			Stats:   e.Stats,
+		}
+		for _, r := range e.Results {
+			key := r.Key()
+			content := authz.RepoContent{
```
Comment (camdencheek) [inline cmd/frontend/graphqlbackend/search_results.go:1945]: @ryanslade My apologies for waiting this long for this comment, but I didn't really read myself in on this whole sub-repo permissions thing until now.

This section is problematic for two reasons:
1) `Key()` isn't _really_ meant to be introspected on. This is my bad for failing to document this, but its intent is to be an opaque, unique key for a search result for deduping/merging results. There are no guarantees to its contents except its use for that purpose.
2) Because of (1), the value of `Path` on the key does not do what you need it to do here. This approach will work alright for `FileMatch` results, but for repo results and commit results, `Path` will be empty. This might be fine for repo results (assuming an empty path is okay), but for commit results, our `CommitMatch` type does not currently contain information about what files the commit modifies.

Looking at the logic in [another PR](https://github.com/sourcegraph/sourcegraph/pull/29771/files) for adding sub-repo perms to listing commits, it appears this information is necessary to do this correctly. 

TBH, I'm not entirely sure what the permission model is for sub-repo filtering of commits. Are users allowed to see commit messages for commits that modify a blocked file? If a commit modifies a blocked file, are users still allowed to see the portion of the diff that doesn't apply to that file?

**Review by camdencheek — CHANGES_REQUESTED** [review]

File: cmd/frontend/graphqlbackend/search_results.go (hunk 4 above)
Comment (rvantonder) [reply cmd/frontend/graphqlbackend/search_results.go:1945]: (I made a low effort attempt to make `key()` private at some point but ran into an issue and gave up. Worth revisiting since this tendency to use `Key` recurred)

Comment (camdencheek) [reply cmd/frontend/graphqlbackend/search_results.go:1945]: I realize I requested changes here without a clear path forward, so feel free to ping me to chat through some options here. I expect it will require some changes to commit/diff search in order for the needed information to be available here.

**Comment by ryanslade** [comment]
Note to self:

We can do a type assertion against search matches so they we don't use the generic key.

For file matches we have a path.
Repo matches can be ignored since existing repo filtering takes care of this. 
Commit matches will be handled by including a set of paths modified by the commit, as mentioned in https://github.com/sourcegraph/sourcegraph/issues/29988

File: cmd/frontend/graphqlbackend/search_results.go (hunk 4 above)
Comment (ryanslade) [reply cmd/frontend/graphqlbackend/search_results.go:1945]: See this note: https://github.com/sourcegraph/sourcegraph/pull/29308#issuecomment-1017771634

**Comment by mollylogue** [comment]
@ryanslade Are you planning to do a follow-up PR with filtering commits? Or are you planning to do that as part of this PR?

**Comment by ryanslade** [comment]
@mollylogue My plan is to merge https://github.com/sourcegraph/sourcegraph/pull/30109 first and then come back to this PR and change it so that it doesn't use `Key()` anymore.

**Comment by ryanslade** [comment]
I've now merged the change that includes modified files in commit matches but after merging those changes as well as changes in `main` there's an issue: 

The aggregator no longer has the `Error` field so I'm not sure what we should do when we encounter an error performing sub-repo filtering. Before, we were able to set the `Error` field so that it could bubble up.

@camdencheek any ideas?

**Comment by camdencheek** [comment]
Aah shoot. That was me. Forgot you were using that. Moving into `doResults` sounds like a good spot to me.

**Comment by tsenart** [comment]
> Aah shoot. That was me. Forgot you were using that. Moving into doResults sounds like a good spot to me.

@camdencheek: Could this be a Job middleware, now that we have this infrastructure?

**Comment by camdencheek** [comment]
@tsenart Yes, absolutely, but I'm not going to ask @ryanslade to rewrite this PR to make that happen. I'll followup with a PR converting this to a job wrapper since I'm actively chipping away at `doResults` and friends anyways

**Comment by ryanslade** [comment]
> @tsenart Yes, absolutely, but I'm not going to ask @ryanslade to rewrite this PR to make that happen. I'll followup with a PR converting this to a job wrapper since I'm actively chipping away at `doResults` and friends anyways

Thanks, appreciate that 😄

**Comment by ryanslade** [comment]
@camdencheek PTAL

I'm about to stop for the day so wanted to give you a chance to give this a look.

The only thing I still want to do is add a couple more cases to `TestApplySubRepoFiltering`

File: cmd/frontend/graphqlbackend/search_results.go [hunk 5]
```diff
@@ -1818,17 +1819,34 @@ func doResults(ctx context.Context, searchInputs *run.SearchInputs, db database.
 		tr.Finish()
 	}()
 
+	errs := &multierror.Error{}
+
 	agg := run.NewAggregator(stream)
```
Comment (camdencheek) [inline cmd/frontend/graphqlbackend/search_results.go:1824]: This is a problematic bit. When we create a new aggregator with a non-nil stream, we don't actually aggregate the results, we stream them as they come, only aggregating stats about the results. This means the `matches` on line 1834 will (usually) be `nil`. Confusing, I know, but we're actively working on making this more intuitive 😄 

I took the liberty of making [a PR](https://github.com/sourcegraph/sourcegraph/pull/30419) against this branch that moves the filter step to a stream func so the results are filtered before they ever hit the aggregator. This way, whether or not the aggregator actually aggregates the results, they will still run through the filter.

File: internal/authz/sub_repo_perms.go [hunk 6]
```diff
@@ -319,6 +313,39 @@ func SubRepoEnabled(checker SubRepoPermissionChecker) bool {
...
+
+	var perms Perms
+	var err error
+	c := RepoContent{
+		Repo: repo,
+	}
+
+	for _, p := range paths {
+		c.Path = p
+		perms, err = checker.Permissions(ctx, a.UID, c)
```
Comment (camdencheek) [inline internal/authz/sub_repo_perms.go:337]: I don't think there is any need to pre-declare perms and err here

```suggestion
	c := RepoContent{
		Repo: repo,
	}

	for _, p := range paths {
		c.Path = p
		perms, err := checker.Permissions(ctx, a.UID, c)
```

File: internal/search/result/file.go [hunk 7]
```diff
@@ -39,7 +38,7 @@ func (f *File) URL() *url.URL {
 // FileMatch represents either:
 // - A collection of symbol results (len(Symbols) > 0)
 // - A collection of text content results (len(LineMatches) > 0)
-// - A result repsenting the whole file (len(Symbols) == 0 && len(LineMatches) == 0)
+// - A result representing the whole file (len(Symbols) == 0 && len(LineMatches) == 0)
```
[the remark this answers is not in the document]
Comment (ryanslade) [reply internal/search/result/file.go:41]: Thanks GoLand for making all the typos jump out at me while reading code :)

**Comment by ryanslade** [comment]
@camdencheek Thanks again for all your assistance with this PR, PTAL
